sentry/src/main/java/io/sentry/SentryTracer.java

```diff
@@ -150,8 +188,9 @@ public void finish() {
 
   @Override
   public void finish(@Nullable SpanStatus status) {
-    if (!root.isFinished()) {
-      root.finish(status);
+    this.finishStatus = status;
+    if (!root.isFinished() && (!waitForChildren || hasAllChildrenFinished())) {
```

Comment:
`root.isFinished()` is an atomic operation, so I'd do the other 2 checks with an inner `if`

Reply:
yep, you're right, no difference, my bad